archive/tar.go

```diff
@@ -648,3 +650,32 @@ func copyBuffered(ctx context.Context, dst io.Writer, src io.Reader) (written in
 	return written, err
 
 }
+
+// hardlinkRootPath returns target linkname, evaluating and bounding any
+// symlink to the parent directory.
+//
+// NOTE: Allow hardlink to the softlink, not the real one. For example,
+//
+//	touch /tmp/zzz
+//	ln -s /tmp/zzz /tmp/xxx
+//	ln /tmp/xxx /tmp/yyy
+//
+// /tmp/yyy should be softlink which be same of /tmp/xxx, not /tmp/zzz.
+func hardlinkRootPath(root, linkname string) (string, error) {
+	ppath, base := filepath.Split(linkname)
+	ppath, err := fs.RootPath(root, ppath)
+	if err != nil {
+		return "", err
+	}
+
+	targetPath := filepath.Join(ppath, base)
+	fi, err := os.Lstat(targetPath)
```

Comment:
Actually, I made it complicated by the `Lstat`. 

As you mentioned, don't allow the `base` to change the `ppath` so that we keep the scope. But it's quite strict. The `base` can be out of the `ppath` but in the `root`, which make the error information accuracy. In this case,  we can check `targetPath` is out of `root` or not. If does, we just return the `root`. 

How about this solution? 😄